aws/resource_aws_qldb_stream.go

```diff
@@ -0,0 +1,349 @@
+					validation.StringLenBetween(1, 32),
+				),
+			},
+
+			"tags": tagsSchema(),
+		},
+	}
+}
+
+func resourceAwsQLDBStreamCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).qldbconn
+
+	// Create the QLDB Stream
+	createOpts := &qldb.StreamJournalToKinesisInput{
+		LedgerName: aws.String(d.Get("ledger_name").(string)),
+		RoleArn:    aws.String(d.Get("role_arn").(string)),
+		StreamName: aws.String(d.Get("stream_name").(string)),
+		Tags:       keyvaluetags.New(d.Get("tags").(map[string]interface{})).IgnoreAws().QldbTags(),
+	}
+
+	if v, ok := d.GetOk("exclusive_end_time"); ok {
+		log.Printf("DEBUG - exclusive_end_time: %#v", v)
+		exclusiveEndTimeValue, _ := time.Parse("2006-01-02T15:04:05Z", v.(string))
+		createOpts.ExclusiveEndTime = &exclusiveEndTimeValue
+	}
+
+	if v, ok := d.GetOk("inclusive_start_time"); ok {
+		log.Printf("DEBUG - inclusive_start_time: %#v", v)
+		inclusiveStartTimeValue, _ := time.Parse("2006-01-02T15:04:05Z", v.(string))
+		createOpts.InclusiveStartTime = &inclusiveStartTimeValue
+	} else {
+		return errors.New("Missing 'inclusive_start_time'")
+	}
+
+	if v, ok := d.GetOk("kinesis_configuration"); ok {
+		createOpts.KinesisConfiguration = &qldb.KinesisConfiguration{}
+
+		values := v.(map[string]interface{})
+		// values := v.(*schema.TypeMap)
+		if value, ok := values["aggregation_enabled"]; ok {
+			aggregationEnabled, err := strconv.ParseBool(value.(string))
+			if err != nil {
+				return errors.New("Error parsing kinesis_configuration.aggregation_enabled")
+			}
+			createOpts.KinesisConfiguration.AggregationEnabled = &aggregationEnabled
+		}
+
+		if value, ok := values["stream_arn"]; ok {
+			streamArn := value.(string)
+			createOpts.KinesisConfiguration.StreamArn = &streamArn
+		}
+	} else if !ok {
+		return errors.New("Missing 'kinesis_configuration'")
+	}
+
+	log.Printf("[DEBUG] QLDB Stream create config: %#v", *createOpts)
+	qldbResp, err := conn.StreamJournalToKinesis(createOpts)
+	if err != nil {
+		return fmt.Errorf("Error creating QLDB Ledger Stream: %s", err)
+	}
+
+	// Set QLDB Stream ID
+	d.SetId(aws.StringValue(qldbResp.StreamId))
+	log.Printf("[INFO] QLDB Ledger Stream Id: %s", d.Id())
+
+	stateConf := &resource.StateChangeConf{
+		Pending:    []string{qldb.StreamStatusImpaired},
+		Target:     []string{qldb.StreamStatusActive},
+		Refresh:    qldbStreamRefreshStatusFunc(conn, d.Get("ledger_name").(string), d.Id()),
+		Timeout:    8 * time.Minute,
+		MinTimeout: 3 * time.Second,
+	}
+
+	_, err = stateConf.WaitForState()
+	if err != nil {
+		return fmt.Errorf("Error waiting for QLDB Ledger status to be \"%s\": %s", qldb.LedgerStateActive, err)
+	}
+
+	// Update our attributes and return
+	return resourceAwsQLDBStreamRead(d, meta)
+}
+
+//TODO: Another edge case to account for?:  Stream resources that are in a terminal state (CANCELED, COMPLETED, and FAILED) are subject to a 7-day retention period. They are automatically hard-deleted after this limit expires.
```

Comment:
Any insight on how best to handle this?
I think there's an edge case here:
- Create a QLDB Stream that expires
- 7 days after it expires it is gone
- `terraform plan` shows this as a missing resource

Ideally this just gets swept under the rug, but I'm not sure the best way to do that.  Some sort of conditional check?  But that means manipulating the state conditionally, which doesn't feel great 😬

Reply:
The usual pattern here is to handle the [terminal states](https://docs.aws.amazon.com/qldb/latest/developerguide/streams.create.html#streams.create.states) as `NotFound`/